gdal_subsetter/transform.py

```diff
@@ -1070,67 +1070,72 @@ def mask_via_combined(self, inputfile, shapefile, outputfile):
+            gdal.SetConfigOption('GDAL_TIFF_INTERNAL_MASK', 'YES')
+            # this changes the maskband data in the dst_dataset
+            dst_dataset.CreateMaskBand(GMF_PER_DATASET)
+
+            # update tmpfile (used as a mask file)
+            with OpenGDAL(tmpfile, GA_Update) as tmp_dataset:
+                for band_index in range(1, tmp_dataset.RasterCount + 1):
+                    self._mask_band(tmp_dataset, band_index, dst_dataset,
+                                    shape_file_layer)
+
+                tmp_dataset.FlushCache()
+
+            dst_dataset.FlushCache()
 
-        # update tmpfile (used as a mask file)
-        tmp_ds = gdal.Open(tmpfile, GA_Update)
-        num = tmp_ds.RasterCount
-
-        # define inner function
-
-        def _mask_band(tmp_ds, band_sn, dst_ds):
-            tmp_band = tmp_ds.GetRasterBand(band_sn)
-            tmp_data = tmp_band.ReadAsArray()
-            tmp_mskband_pre = tmp_band.GetMaskBand()
-            tmp_msk_pre = tmp_mskband_pre.ReadAsArray()
-            tmp_nodata_pre = tmp_band.GetNoDataValue()
-            np_dt = gdal_array.GDALTypeCodeToNumericTypeCode(tmp_band.DataType)
-            tmp_band.WriteArray(np.zeros(tmp_data.shape, np_dt))
-            # this flushCache() changes the all values in the maskband data to 255.
-            tmp_band.FlushCache()
-            bands = [band_sn]
-            burn_value = 1
-            burns = [burn_value]
-            # following statement modify the maskband data to 255.
-            # The original maskband data is stored in tmp_msk_pre.
-            err = gdal.RasterizeLayer(tmp_ds, bands, ly, burn_values=burns,
-                                      options=['ALL_TOUCHED=TRUE'])
-            tmp_ds.FlushCache()
-            # combine original tmp mask band with tmp_data
-            tmp_band = tmp_ds.GetRasterBand(band_sn)
-            # tmp_data includes 0 and 1 values, where 0 indicates no valid pixels, and 1 indicates valid pixels.
-            tmp_data = tmp_band.ReadAsArray()
-            out_band = dst_ds.GetRasterBand(band_sn)
-            out_data = out_band.ReadAsArray()
-            # set out_band with nodata value if the tmp_band has nodata value
-            if tmp_nodata_pre:
-                msk = tmp_data == 0
-                out_data[msk] = tmp_nodata_pre
-                out_band.WriteArray(out_data)
-                out_band.SetNoDataValue(tmp_nodata_pre)
-
-            # modify out_mskband
-            out_mskband = out_band.GetMaskBand()
-            out_msk = tmp_data*tmp_msk_pre
-            out_mskband.WriteArray(out_msk)
-            out_mskband.FlushCache()
-            out_band.FlushCache()
-
-        for band_sn in range(1, num + 1):
-            _mask_band(tmp_ds, band_sn, dst_ds)
-
-        dst_ds.FlushCache()
-        tmp_ds.FlushCache()
-        dst_ds = None
-        tmp_ds = None
         return outputfile
 
+    def _mask_band(self, tmp_ds: gdal.Dataset, band_index: int,
```

Comment:
This probably needs some more thorough attention (there's a lot in here), but migrating `_mask_band` out of the `HarmonyAdapter.mask_via_combined` method makes that `HarmonyAdapter.mask_via_combined` much more readable.

Reply:
There's a couple of other places that do the same thing (`_process_projected` and `_process_geogcs` inside `HarmonyAdapter.geotiff2netcdf_direct`). The one thing those functions all have in common is that they are only used by the method that they are defined within. I guess it might have been some way of indicating that the functionality was internal to the parent class method, but (personal opinion) it mainly just made those class methods a little more difficult to grok.

I will play with those other two functions at some point (I wonder if they need to be separate or could be combined, anyway)